pkg/utils/symbol_table.go

```diff
@@ -0,0 +1,212 @@
+package utils
+
+import (
+	"errors"
+	"sort"
+	"sync"
+)
+
+// The Symbol interface defines what is needed from a symbol implementation in
+// order to facilitate the lookup functionalities provided by SymbolTable.
+// Implementations of Symbol can hold various types of information relevant to
+// the type of symbol they represent.
+type Symbol interface {
+	Name() string
+	Address() uint64
+	Contains(address uint64) bool
+}
+
+// SymbolTable is used to hold information about symbols (mapping from symbolic
+// names used in code to their address) in a certain executable.
+// It can be used to hold symbols from an ELF binary, or symbols of the entire
+// kernel and its modules.
+// It provides functions to lookup symbols by address and name.
+type SymbolTable[T Symbol] struct {
+	mu sync.RWMutex
+	// All symbols sorted by their address in descending order,
+	// for quick binary searches by address.
+	sortedSymbols []*T
+	// If lazyNameLookup is true, the symbolsByName map
+	// will be populated only when a failed lookup occurs.
+	symbolsByName  map[string][]*T
+	lazyNameLookup bool
+}
+
+var ErrSymbolNotFound = errors.New("symbol not found")
+
+// Creates a new SymbolTable. If lazyNameLookup is true, the mapping from
+// name to symbol will be populated only when a failed lookup occurs.
+// This reduces memory footprint at the cost of the time it takes to lookup
+// a symbol name for the first time.
+func NewSymbolTable[T Symbol](lazyNameLookup bool) *SymbolTable[T] {
+	return &SymbolTable[T]{
+		sortedSymbols:  make([]*T, 0),
+		symbolsByName:  make(map[string][]*T),
+		lazyNameLookup: lazyNameLookup,
+	}
+}
+
+// Adds a slice of symbols to the symbol table.
+func (st *SymbolTable[T]) AddSymbols(symbols []*T) {
+	st.mu.Lock()
+	defer st.mu.Unlock()
+
+	// Add the new symbols to the sorted slice (which now becomes unsorted).
+	// Allocate the slice with the needed capacity to avoid overallocation.
+	oldSymbols := st.sortedSymbols
+	newLen := len(oldSymbols) + len(symbols)
+	st.sortedSymbols = make([]*T, 0, newLen)
+	st.sortedSymbols = append(st.sortedSymbols, oldSymbols...)
+
+	// If lazyNameLookup is false, we update the name to symbol mapping for
+	// each new symbol
+	if !st.lazyNameLookup {
+		for _, symbol := range symbols {
+			st.sortedSymbols = append(st.sortedSymbols, symbol)
```

Comment:
This is redundant, you could've just as well copied here the same way as the `else` section, which would be more time efficient. 
Note how you're not relying on `st.sortedSymbols` at all in the rest of the block. So you can actually skip the `if..else` copy first like in the `else`, then only do this loop in a single `if`.